src/main/java/com/redhat/devtools/lsp4ij/features/foldingRange/LSPFoldingRangeBuilder.java

```diff
@@ -54,8 +66,30 @@ protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> descrip
         }
 
         PsiFile file = root.getContainingFile();
+        List<FoldingRange> foldingRanges = getFoldingRanges(file);
+        if (!ContainerUtil.isEmpty(foldingRanges)) {
+            for (FoldingRange foldingRange : foldingRanges) {
+                TextRange textRange = getTextRange(foldingRange, document);
+                if (textRange != null) {
+                    String collapsedText = foldingRange.getCollapsedText();
+                    if (collapsedText != null) {
+                        descriptors.add(new FoldingDescriptor(root.getNode(), textRange, null, collapsedText));
+                    } else {
+                        descriptors.add(new FoldingDescriptor(root.getNode(), textRange));
+                    }
+                }
+            }
+        }
+    }
+
+    @NotNull
+    static List<FoldingRange> getFoldingRanges(@Nullable PsiFile file) {

```

Comment:
Note that all of the reusable aspects are package-protected so that they don't become part of the public API of LSP4IJ but are usable by the code block provider.